File: src/backend/services/web.admin/dashboard-app/src/lib/types/catalog.ts
import { z } from 'zod';

export const CatalogItemScheme = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string(),
  price: z.number(),
  image: z.string(),
  currency: z.string()
});

export const CatalogItemsScheme = z.object({
  catalog_items: z.array(CatalogItemScheme),
  total: z.number(),
  offset: z.number(),
  limit: z.number()
});

export const CategoriesScheme = z.array(
  z.object({
    id: z.number(),
    name: z.string()
  })
);

export type CatalogItem = z.infer<typeof CatalogItemScheme>;
export type CatalogItems = z.infer<typeof CatalogItemsScheme>;
export type Categories = z.infer<typeof CategoriesScheme>;

File: src/backend/services/catalog-api/src/models/catalog.rs
use crate::schema::catalog;
use bigdecimal::BigDecimal;
use rocket::serde::{Deserialize, Serialize};
use chrono;
use rocket_okapi::{JsonSchema};

#[derive(Debug, Clone, PartialEq, Queryable, AsChangeset)]
#[diesel(table_name = catalog)]
pub struct Catalog {
    pub id: i32,
    pub name: String,
    pub description: Option<String>,
    pub image: String,
    pub price: BigDecimal,
    pub currency: String,
    pub category: String,
    pub created_at: chrono::NaiveDateTime,
    pub updated_at: chrono::NaiveDateTime,
}

#[derive(Debug, Clone, AsChangeset)]
#[diesel(table_name = catalog)]
pub struct UpdateCatalog {
    pub name: String,
    pub description: Option<String>,
    pub image: String,
    pub price: BigDecimal,
    pub currency: String,
    pub category: Option<String>,
    pub updated_at: chrono::NaiveDateTime,
}

#[derive(Insertable)]
#[diesel(table_name = catalog)]
pub struct NewCatalog {
    pub name: String,
    pub description: Option<String>,
    pub image: String,
    pub price: BigDecimal,
    pub currency: String,
    pub category: Option<String>,
    pub created_at: chrono::NaiveDateTime,
    pub updated_at: chrono::NaiveDateTime,
}

#[derive(Serialize, Deserialize, Clone, JsonSchema)]
#[serde(crate = "rocket::serde")]
pub struct CatalogRequest {
    pub name: String,
    pub description: Option<String>,
    pub image: String,
    pub price: f64,
    pub currency: String,
    pub category: Option<String>,
}

#[derive(Serialize, Deserialize, JsonSchema)]
#[serde(crate = "rocket::serde")]
pub struct CatalogItem {
    pub id: i32,
    pub name: String,
    pub description: Option<String>,
    pub image: String,
    pub price: f64,
    pub currency: String,
    pub category: String,
}

#[derive(Serialize, Deserialize, JsonSchema)]
#[serde(crate = "rocket::serde")]
pub struct CatalogsResponse {
    pub catalog_items: Vec<CatalogItem>,
    pub total: i64,
    pub offset: i64,
    pub limit: i64,
}
